### ISSUE TITLE
Multiple errors when running on CDM 5.4 in an Oracle DB

### ISSUE THREAD

**Issue by tlecarrour-ee** [OHDSI/DataQualityDashboard#433]
Hi,

For the first time, I'm running DQD (v2.1.2) on a CDM (v5.4) in an Oracle database. I end up with **a lot** of errors (200+) and no JSON result file?!

I've looked into some of those errors and I have to admit that I'm a bit puzzled. Here are 2 examples…

1) `FIELD_standardConceptRecordCompleteness_VISIT_OCCURRENCE_DISCHARGE_TO_CONCEPT_ID.txt`

The error is `java.sql.SQLSyntaxErrorException: ORA-00904: "CDMTABLE"."DISCHARGE_TO_CONCEPT_ID": invalid identifier`.

Indeed, this field `DISCHARGE_TO_CONCEPT_ID` does not exist in CDM v5.4, where it has been deprecated in favor of `DISCHARGED_TO_CONCEPT_ID` (with a `D`). Is this a "normal" error that is ignored in the final report?

2) `FIELD_plausibleValueLow_DRUG_EXPOSURE_DRUG_EXPOSURE_END_DATE.txt`

The error is `ORA-01858: a non-numeric character was found where a numeric was expected`.

The SQL is:

```sql
SELECT num_violated_rows,
  CASE
    WHEN denominator.num_rows = 0 THEN 0
    ELSE 1.0*num_violated_rows/denominator.num_rows
  END AS pct_violated_rows,
  denominator.num_rows AS num_denominator_rows
FROM (SELECT COUNT(violated_rows.violating_field) AS num_violated_rows
	FROM (SELECT 'DRUG_EXPOSURE.DRUG_EXPOSURE_END_DATE' AS violating_field,
		  cdmTable.*
		FROM omop_cdm.DRUG_EXPOSURE cdmTable
      WHERE TO_DATE(cdmTable.DRUG_EXPOSURE_END_DATE , 'YYYYMMDD') < TO_DATE('1950-01-01' , 'YYYYMMDD')
		/*violatedRowsEnd*/
	 ) violated_rows
 ) violated_row_count,
(SELECT COUNT(*) AS num_rows
	FROM omop_cdm.DRUG_EXPOSURE cdmTable
	  WHERE DRUG_EXPOSURE_END_DATE IS NOT NULL
 ) denominator
 ;
```

The origin of the error is that `cdmTable.DRUG_EXPOSURE_END_DATE` is already a date and should not be passed to `TO_DATE`. **But**, if I remove this first `TO_DATE` I run into another error, for `YYYYMMDD` is not the right format for '1950-01-01' , but `YYYY-MM-DD`.

---

So, my questions are : Is running DQD on Oracle really different from running it on PostgreSQL?! Am I missing something really obvious?!

**Comment by katy-sadowski** [comment]
Hi there!

For #1 - are you sure you've specified cdmVersion = "5.4" when running DQD?  In the 5.4 threshold files the column name you're referring to does have `discharged_` with the correct spelling

For #2 - this actually seems like a SqlRender issue.  The SQL queries in DQD get translated into the various supported SQL dialects using SqlRender.  For Oracle SQL, seems we need to add a cast to string for the initial param in TO_DATE and fix the date format in the second param.  Would you mind filing an issue forthat in the SqlRender repo [here](https://github.com/OHDSI/SqlRender/issues)? Thanks!

**Comment by tlecarrour-ee** [comment]
Hi @katy-sadowski! Thanks for you reply!

> For #1 - are you sure you've specified cdmVersion = "5.4" when running DQD?

I had no clue one had to provide the CDM version when running DQD! Shame on me! From the function's signature, it's obvious that it defaults to `5.3`. My bad! Sorry!

Now I have **A LOT** of `ORA-01000: maximum open cursors exceeded` that I'll have to investigate. But, I guess, it's a totally different problem.

> For #2 - this actually seems like a SqlRender issue.

I've reported it as [issue 327](https://github.com/OHDSI/SqlRender/issues/327).

**Comment by katy-sadowski** [comment]
@schuemie I looked back in the git blame and seems the casts were added [here](https://github.com/OHDSI/DataQualityDashboard/pull/88/commits/95196f1a045dfa094d0077a7ca6490b011ddbcb3) due to a BigQuery [issue](https://github.com/OHDSI/DataQualityDashboard/issues/98).  

I _think_ we could handle both issues by conditionally comparing the un-casted field to a date vs datetime depending on the field being checked.  I'll work on this and report back 👌

**Comment by schuemie** [comment]
Thanks @katy-sadowski for digging into this. I don't think a conditional casting will help here though. The problem is that in SQL Server (and therefore in OhsdiSql), `CAST(... AS DATE)` can be used both to cast DATETIME fields to date, as well as VARCHAR fields. The current [Oracle translation rule](https://github.com/OHDSI/SqlRender/blob/main/inst/csv/replacementPatterns.csv#L4) only covers the VARCHAR case, which would mean you would still have a problem if you only applied it when the field is a DATETIME.

Let me see if I can work some magic on the SqlRender side.

**Comment by schuemie** [comment]
It seems this specific error is solved by [this SqlRender commit](https://github.com/OHDSI/SqlRender/commit/edaf82e344e42c35d5f53ab83950dce21a585a05). 

When I run DQD on Oracle (CDM V5.4) I do still see several 'ORA-00932: inconsistent datatypes: expected - got CLOB' error messages. These arise when you call `COUNT()` on a field that has only NULL values, which actually seems like an Oracle bug to me (that I don't have a good workaround for).

**Comment by katy-sadowski** [comment]
> I don't think a conditional casting will help here though. The problem is that in SQL Server (and therefore in OhsdiSql), `CAST(... AS DATE)` can be used both to cast DATETIME fields to DATE, as well as VARCHAR fields to DATE. The current [Oracle translation rule](https://github.com/OHDSI/SqlRender/blob/main/inst/csv/replacementPatterns.csv#L4) only covers the VARCHAR case, which would mean you would still have a problem if you only applied it when the field is a DATETIME.

I was thinking not to cast the field at all, and compare either to a datetime or a date (i.e. '1950-01-01 00:00:00' vs '1950-01-01') depending on the type of the field. However that was certainly not a sure bet...so I'm glad to see a SqlRender fix in the works!

What are your thoughts on the second issue @tlecarrour-ee mentioned here:

> if I remove this first TO_DATE I run into another error, for YYYYMMDD is not the right format for '1950-01-01' , but YYYY-MM-DD.

Does this also need to be addressed in SqlRender?

**Comment by schuemie** [comment]
YYYYMMDD is the standard format supported by SqlRender. This seems to be fine now, right? (I can run DQD on Oracle, and only see the error I mentioned related to `COUNT()` on `NULL` fields.)

**Comment by tlecarrour-ee** [comment]
Hi @schuemie,
This morning I ran DQD with the `develop` versions of SqlRender and DQD.
The good news is I now have a result file! The bad news is I still have a lot of errors. Part of them are related to #347, the rest to… the date format problem I mentioned earlier.

For instance, in `FIELD_plausibleValueLow_VISIT_OCCURRENCE_VISIT_START_DATE.txt` I get a `ORA-01843: not a valid month`. I can re-run the query in Oracle and get the same error. When I change the format from `YYYYMMDD` to `YYYY-MM-DD` in the query mentionned in this `txt` file, I get a result for the query.

**Comment by schuemie** [comment]
The solution here would be to always use the `YYYYMMDD` format in DQD, which should work across all SQL dialects. Where exactly does DQD supply the date in `YYYY-MM-DD` format?

**Comment by katy-sadowski** [comment]
Sounds good @schuemie !  Those dates come from the threshold files.  I can convert them to `YYYYMMDD`.

**Comment by katy-sadowski** [comment]
Date formats are fixed in the v2.2.0 release.
